### PR TITLE
[WFCORE-482] Support the log4j2 API delegating log messages to the JB…

### PR DESCRIPTION
**Pull request by jamezp**
…oss Log Manager.

https://issues.redhat.com/browse/WFCORE-482

### CODE REVIEW

File: logging/WFCORE-482-log4j2-support.adoc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+== Requirements
+
+=== Hard Requirements
+
+* A user can use a log4j2 logger, `org.apache.logging.log4j.Logger`, which will pass the log message to the JBoss Log
+  Manager which is the log manager WildFly uses.
+* The log4j-api and implementation will be added to all deployments by default.
+* The log4j2 loggers created should honor the JBoss Log Manager's log context associated with the deployment.
+** This means if the deployment includes a logging configuration for a supported log manager the logger created should
```
Comment (tommaso-borgato) [inline logging/WFCORE-482-log4j2-support.adoc:53]: @jamezp same as above.... I'd some more info to understand this

File: logging/WFCORE-482-log4j2-support.adoc [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+   honor that log context.
+* The `ThreadContext.getContext()` will be copied to `MDC` in for the log record.
```
Comment (tommaso-borgato) [inline logging/WFCORE-482-log4j2-support.adoc:55]: @jamezp same as above.... I'd some more info to understand this and its implications

File: logging/WFCORE-482-log4j2-support.adoc [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+* The `ThreadContext.getImmutableStack()` will be treated as `NDC` and formatted as such for the log record.
```
Comment (tommaso-borgato) [inline logging/WFCORE-482-log4j2-support.adoc:56]: @jamezp same as above

File: logging/WFCORE-482-log4j2-support.adoc [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+=== Nice-to-Have Requirements
+
+* It would be nice if the deployment already includes the log4j2 API and log4j2 core libraries, then the API should not
```
Comment (tommaso-borgato) [inline logging/WFCORE-482-log4j2-support.adoc:60]: @jamezp are there specific consequences that's worth mentioning?

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: logging/WFCORE-482-log4j2-support.adoc [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1352[EAP7-1352]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
```
Comment (tommaso-borgato) [inline logging/WFCORE-482-log4j2-support.adoc:31]: @jamezp can you please add tborgato@redhat.com ?

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: logging/WFCORE-482-log4j2-support.adoc [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+* The log4j2 loggers created should honor the JBoss Log Manager's log context associated with the deployment.
+** This means if the deployment includes a logging configuration for a supported log manager the logger created should
+   honor that log context.
+* The `ThreadContext.getContext()` will be copied to `MDC` in for the log record.
+* The `ThreadContext.getImmutableStack()` will be treated as `NDC` and formatted as such for the log record.
+
+=== Nice-to-Have Requirements
+
+* It would be nice if the deployment already includes the log4j2 API and log4j2 core libraries, then the API should not
+  automatically be added to the deployment.
```
Comment (tommaso-borgato) [inline logging/WFCORE-482-log4j2-support.adoc:61]: @jamezp are there specific consequences that's worth mentioning?

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: logging/WFCORE-482-log4j2-support.adoc [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+* Any consumer of WildFly Core that uses the logging subsystem.
+
+== Requirements
+
+=== Hard Requirements
+
+* A user can use a log4j2 logger, `org.apache.logging.log4j.Logger`, which will pass the log message to the JBoss Log
+  Manager which is the log manager WildFly uses.
+* The log4j-api and implementation will be added to all deployments by default.
```
[the remark this answers is not in the document]
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:51]: > * `implementation` is the separate community project https://github.com/jboss-logging/log4j2-jboss-logmanager, right?

That is correct. It's a separate project which implements the `org.apache.logging.log4j:log4j-api`.

> * IUUC the paragraph [Admin_Guide.html#add-logging-api-dependencies](https://docs.wildfly.org/21/Admin_Guide.html#add-logging-api-dependencies) _, "adding `log4j-api` and implementation to all deployments by default"_ is the same strategy adopted for supporting the other already supported logging frameworks (slf4j, commons-logging, jboss-logging), right ?

Yes that is correct as well.

File: logging/WFCORE-482-log4j2-support.adoc (hunk 1 above)
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:53]: By supported log manager this means only a log4j 1.x or JBoss Log Manager configuration file. slf4j, commons-logging and jboss-logging are only logging facades so they do not support any specific configuration files.

File: logging/WFCORE-482-log4j2-support.adoc (hunk 2 above)
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:55]: I'll see if I can word this better. Basically log4j2 replaces NDC and MDC with `ThreadContext`. So we'll copy the `ThreadContext.getContext()` to get a copy of the MDC and we use the `ThreadContext.getImmutableStack()` to create an NDC string for the logrecord.

File: logging/WFCORE-482-log4j2-support.adoc (hunk 3 above)
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:56]: Oops got ahead of myself, but it's answered above as well :)

File: logging/WFCORE-482-log4j2-support.adoc (hunk 6 above)
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:61]: I think I'll delete this. I did some investigation on this and it never really felt right. The idea was if the user had log4j2-core in their deployment we wouldn't use our own implementation. However it wasn't really a clean design and seemed flawed to me so I don't think, at least at this point, we should support it.

File: logging/WFCORE-482-log4j2-support.adoc (hunk 2 above)
Comment (tommaso-borgato) [reply logging/WFCORE-482-log4j2-support.adoc:57]: @jamezp IUUC this is just the technical explanation on how that information (namely `ThreadContext`) is mapped from log4j2 APIs onto the underlying implementation, right?

File: logging/WFCORE-482-log4j2-support.adoc (hunk 3 above)
Comment (tommaso-borgato) [reply logging/WFCORE-482-log4j2-support.adoc:58]: @jamezp yet another technical explanation on how `log4j2` info is mapped, right?

File: logging/WFCORE-482-log4j2-support.adoc (hunk 2 above)
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:57]: Yes. I guess more of a note that the previous MDC and NDC objects are now `ThreadContext` and how they'll be mapped.

File: logging/WFCORE-482-log4j2-support.adoc (hunk 3 above)
Comment (jamezp) [reply logging/WFCORE-482-log4j2-support.adoc:58]: Correct.

**Review by tommaso-borgato — APPROVED** [review]

**Pull request merged**